### PR TITLE
Feature/bootstrap

### PR DESCRIPTION
**Pull request by eeerin**
[SP-2770](https://structure.atlassian.net/browse/SP-2770)

### CODE REVIEW

**Comment by eeerin** [comment]
moving this to work in progress, because Brandon wants me to stop the bootstrap if they have any workflows, views or endpoints. I'll add that tomorrow.

**Comment by eeerin** [comment]
This now will not run the bootstrap command, if it has already been skipped or completed.

File: commands/configure/index.js [hunk 1]
```diff
@@ -108,6 +106,15 @@ program
       const downloaded = await experienceDownload(null, {}, loadedConfig);
       if (downloaded) {
         logResult('success', 'Downloaded all experience resources!', 'green');
+      } else {
+        const { shouldBootstrap } = await inquirer.prompt([{
```
Comment (spacetc62) [inline commands/configure/index.js]: before you prompt, check the FTUE. if they have already done or skipped bootstrapping, don't ask.

File: commands/configure/index.js [hunk 2]
```diff
@@ -108,6 +106,15 @@ program
...
+          type: 'confirm',
+          name: 'shouldBootstrap',
+          message: `Do you want to bootstrap your experience for application ${appInfo.name}?`
+        }]);
+        if (shouldBootstrap) {
+          await experienceBootStrap({}, loadedConfig, appInfo);
```
Comment (spacetc62) [inline commands/configure/index.js]: really? BootStrap? :P

File: lib/experience-bootstrap.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  const api = await getApi({ apiToken });
+  try {
+    applicationInfo = applicationInfo || await api.application.get({ applicationId });
+    const bootstrap = hasBootstrapped(applicationInfo);
+    if (bootstrap) {
+      return logResult('Cannot Complete', `Bootstrapping has already been ${bootstrap.status} for ${applicationName}`, 'yellow');
+    }
+    const results = await api.experience.bootstrap({ applicationId });
+    logResult('Bootstrap Username', `${results.email}`);
+    logResult('Bootstrap Password', `${results.password}`);
```
Comment (spacetc62) [inline lib/experience-bootstrap.js]: also print out the URL they can go to to view this stuff

Comment (spacetc62) [reply lib/experience-bootstrap.js]: also, I'd probably move this print out *below* the download, so it is more obvious

File: commands/configure/index.js (hunk 1 above)
Comment (spacetc62) [reply commands/configure/index.js]: also, if they say no, you should then set the FTUE flag on the application to skipped, just like the frontend

File: lib/experience-bootstrap.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  loadConfig,
+  logResult,
+  logError
+} = require('./utils');
+const getApi = require('./get-api');
+const { experience } = require('./get-download-params');
+const getDownloader = require('./get-downloader');
+const downloader = getDownloader(experience);
+const { pipe, propOr, find, propEq } = require('omnibelt');
+const hasBootstrapped = pipe(propOr([], 'ftueTracking'), find(propEq('name', 'experience')));
```
Comment (spacetc62) [inline lib/experience-bootstrap.js]: should only be considered "bootstrapped" if `version >= 2` - see equivalent front end check (in `shouldShowModal`)

File: commands/configure/index.js [hunk 5]
```diff
@@ -108,6 +106,27 @@ program
...
+          } else {
+            if (!appInfo.ftueTracking) { // eslint-disable-line max-depth
+              appInfo.ftueTracking = [];
+            }
+            appInfo.ftueTracking.push({
+              name: 'experience',
+              version: 3,
+              status: 'skipped'
+            });
+            await api.application.update({ appilcationId: appInfo.id }, { appInfo });
```
Comment (spacetc62) [inline commands/configure/index.js]: typo

Comment (spacetc62) [reply commands/configure/index.js]: also, this is 'patch' not 'update'. And I'd only update ftueTracking, don't patch with the rest of the info

File: commands/configure/index.js [hunk 6]
```diff
@@ -108,6 +106,27 @@ program
...
+        if (!hasBootstrapped(appInfo)) {
+          const { shouldBootstrap } = await inquirer.prompt([{
+            type: 'confirm',
+            name: 'shouldBootstrap',
+            message: `Do you want to bootstrap your experience for application ${appInfo.name}?`
+          }]);
+          if (shouldBootstrap) {
+            await experienceBootstrap({}, loadedConfig, appInfo);
+          } else {
+            if (!appInfo.ftueTracking) { // eslint-disable-line max-depth
```
Comment (spacetc62) [inline commands/configure/index.js]: refactor to remove this linter warning

File: lib/experience-bootstrap.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  const api = await getApi({ apiToken });
+  try {
+    applicationInfo = applicationInfo || await api.application.get({ applicationId });
+    const bootstrap = hasBootstrapped(applicationInfo);
+    if (bootstrap) {
+      return logResult('Cannot Complete', `Bootstrapping has already been ${bootstrap.status} for ${applicationName}`, 'yellow');
+    }
+    const results = await api.experience.bootstrap({ applicationId });
+    await downloader(null, {}, { apiToken, applicationId });
+    logResult('Experience URL', `${applicationInfo.endpointSlug}`);
```
Comment (spacetc62) [inline lib/experience-bootstrap.js]: this is not the URL, this is just the slug - you need to build the full URL they can copy paste

File: lib/utils.js [hunk 8]
```diff
@@ -355,6 +356,12 @@ const utils = {
     }
     command.type = newType;
     return true;
+  },
+  hasBootstrapped: (application) => {
+    const result = find(propEq('name', 'experience'), propOr([], 'ftueTracking', application));
+    if (result && result.version >= 2) {
```
[the remark this answers is not in the document]
Comment (spacetc62) [reply lib/utils.js]: nevermind, you use the FTUE entry

File: lib/experience-bootstrap.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  const api = await getApi({ apiToken });
+  try {
+    applicationInfo = applicationInfo || await api.application.get({ applicationId });
+    const bootstrap = hasBootstrapped(applicationInfo);
+    if (bootstrap) {
+      return logResult('Cannot Complete', `Bootstrapping has already been ${bootstrap.status} for ${applicationName}`, 'yellow');
+    }
+    const results = await api.experience.bootstrap({ applicationId });
+    await downloader(null, {}, { apiToken, applicationId });
+    logResult('Experience URL', `https://${applicationInfo.endpointSlug}.onlosant.com`);
```
Comment (spacetc62) [inline lib/experience-bootstrap.js]: you need to tack on the `resourceSuffix` here if one is provided in `results`

File: lib/utils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (eeerin) [reply lib/utils.js]: yeah I thought it would be nice to tell them why they cant do a bootstrap, either "skipped" or "completed" 😄

**Pull request merged**